megatron/core/dist_checkpointing/strategies/filesystem_async.py

```diff
@@ -366,45 +367,74 @@ def write_preloaded_data(
         mem_before = _process_memory()
         use_msc = kwargs.get("use_msc", False)
 
+        # Retry configuration
+        max_retries = kwargs.get("max_ckpt_save_retries", 3)
+        retry_delay = kwargs.get("ckpt_save_retry_delay", 10.0)  # seconds
+
         local_results = []
-        try:
-            file_name, storage_key, (bytes_data, tensor_data) = write_bucket
-            extra_kwargs = {}
-            if "serialization_format" in inspect.signature(_write_item).parameters:
-                from torch.distributed.checkpoint.filesystem import SerializationFormat
-
-                extra_kwargs["serialization_format"] = SerializationFormat.TORCH_SAVE
-            if use_msc:
-                import multistorageclient as msc
-
-                open_file = msc.open
-            else:
-                open_file = open
-            with open_file(file_name, "wb") as stream:
-                for write_item, data in bytes_data:
-                    local_results.append(
-                        _write_item(
-                            *transform_list, stream, data, write_item, storage_key, **extra_kwargs
+        local_output = None
+
+        for attempt in range(max_retries):
+            try:
+                file_name, storage_key, (bytes_data, tensor_data) = write_bucket
+                extra_kwargs = {}
+                if "serialization_format" in inspect.signature(_write_item).parameters:
+                    from torch.distributed.checkpoint.filesystem import SerializationFormat
+
+                    extra_kwargs["serialization_format"] = SerializationFormat.TORCH_SAVE
+                if use_msc:
+                    import multistorageclient as msc
+
+                    open_file = msc.open
+                else:
+                    open_file = open
+
+                # Reset results for each retry attempt
+                local_results = []
+
+                with open_file(file_name, "wb") as stream:
```

Comment:
Instead of trying to write the whole file every attempt, Can you use io.ByteIO in some number of buffers and apply retrying the write and fsync on each buffer? 
`_write_item` calls torch.save which computes CRC32 of the tensor and do write. 
So, if we retry the whole steps, then we will do this computation again. 

What we  want to achieve here is to have resiliency in the filesystem backend.

Reply:
Ack